c3.php

```diff
@@ -8,399 +8,442 @@
 
-// Load Codeception Config
-$configDistFile = realpath(__DIR__) . DIRECTORY_SEPARATOR . 'codeception.dist.yml';
-$configFile = realpath(__DIR__) . DIRECTORY_SEPARATOR . 'codeception.yml';
+            // fix for improperly encoded JSON in Code Coverage cookie with WebDriver.
+            // @see https://github.com/Codeception/Codeception/issues/874
+            if (!is_array($cookie)) {
+                $cookie = json_decode($cookie, true);
+            }
 
-if (isset($_SERVER['HTTP_X_CODECEPTION_CODECOVERAGE_CONFIG'])) {
-    $configFile = realpath(__DIR__) . DIRECTORY_SEPARATOR . $_SERVER['HTTP_X_CODECEPTION_CODECOVERAGE_CONFIG'];
-}
-if (file_exists($configFile)) {
-    // Use codeception.yml for configuration.
-} elseif (file_exists($configDistFile)) {
-    // Use codeception.dist.yml for configuration.
-    $configFile = $configDistFile;
-} else {
-    __c3_error(sprintf("Codeception config file '%s' not found", $configFile));
-}
-try {
-    \Codeception\Configuration::config($configFile);
-} catch (\Exception $e) {
-    __c3_error($e->getMessage());
-}
+            if ($cookie) {
+                foreach ($cookie as $key => $value) {
+                    if (!empty($value)) {
+                        $_SERVER["HTTP_X_CODECEPTION_" . strtoupper($key)] = $value;
+                    }
+                }
+            }
+        }
 
-if (!defined('C3_CODECOVERAGE_MEDIATE_STORAGE')) {
+        if (!array_key_exists('HTTP_X_CODECEPTION_CODECOVERAGE', $_SERVER)) {
+            return;
+        }
 
-    // workaround for 'zend_mm_heap corrupted' problem
-    gc_disable();
+        if (!function_exists('__c3_error')) {
+            function __c3_error($message)
+            {
+                $errorLogFile = defined('C3_CODECOVERAGE_ERROR_LOG_FILE') ?
+                    C3_CODECOVERAGE_ERROR_LOG_FILE :
+                    C3_CODECOVERAGE_MEDIATE_STORAGE . DIRECTORY_SEPARATOR . 'error.txt';
+                if (is_writable($errorLogFile)) {
+                    file_put_contents($errorLogFile, $message);
+                } else {
+                    $message = "Could not write error to log file ($errorLogFile), original message: $message";
+                }
+                if (!headers_sent()) {
+                    header('X-Codeception-CodeCoverage-Error: ' . str_replace("\n", ' ', $message), true, 500);
+                }
+                setcookie('CODECEPTION_CODECOVERAGE_ERROR', $message);
+            }
+        }
 
-    $memoryLimit = ini_get('memory_limit');
-    $requiredMemory = '384M';
-    if ((substr($memoryLimit, -1) === 'M' && (int)$memoryLimit < (int)$requiredMemory)
-        || (substr($memoryLimit, -1) === 'K' && (int)$memoryLimit < (int)$requiredMemory * 1024)
-        || (ctype_digit($memoryLimit) && (int)$memoryLimit < (int)$requiredMemory * 1024 * 1024)
-    ) {
-        ini_set('memory_limit', $requiredMemory);
-    }
+        // Autoload Codeception classes
```

Comment:
This is a new major version that requires composer, so it would be better to make it depend on Codeception ^5.0 and remove this require block because functions will be loaded by Composer automatically.

Update: If codecept.phar file is used, we have to require codecept.phar/vendor/autoload.php file instead.

Reply:
Not using Composer autoloader and directly including the file should still work, for example:

```
require __DIR__ . '/vendor/codeception/c3/c3.php';

$c3 = new Codeception\C3(__DIR__); // pass the directory that contains Codeception config files
$c3->run();
```

Reply:
Or do you mean that since this will be a new major version of c3, more legacy loading methods can be removed?

Also - I don't know when Codeception v5 will be out, but it's probably better if we can keep support for Codeception v4 as well?